qiskit/transpiler/passes/optimization/template_matching/template_substitution.py

```diff
@@ -492,30 +502,76 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
         from sympy.parsing.sympy_parser import parse_expr
 
         circuit_params, template_params = [], []
+        circ_param_str = ""
+        sub_params = {}
 
         template_dag_dep = copy.deepcopy(self.template_dag_dep)
 
+        # add parameters from circuit to circuit_params, as well as a
+        # string containing parameters to check for duplicates from the template
         for idx, t_idx in enumerate(template_sublist):
             qc_idx = circuit_sublist[idx]
             circuit_params += self.circuit_dag_dep.get_node(qc_idx).op.params
-            template_params += template_dag_dep.get_node(t_idx).op.params
+            for param_exp in self.circuit_dag_dep.get_node(qc_idx).op.params:
+                if isinstance(param_exp, ParameterExpression):
+                    circ_param_str += str(param_exp)
+
+        # add parameters from template to template_params, as well as replace
+        # node parameters that have duplicate names in the circuit params
+        # create a dict of sub_params to substitute remaining nodes not in
+        # the template sublist
+        for idx, t_idx in enumerate(template_sublist):
+            node = template_dag_dep.get_node(t_idx)
+            sub_node_params = []
+            for t_param_exp in node.op.params:
+                if isinstance(t_param_exp, ParameterExpression):
+                    for t_param in t_param_exp.parameters:
+                        if t_param.name in circ_param_str:
+                            new_param_name = "".join(
+                                random.choice(string.ascii_lowercase) for i in range(8)
+                            )
+                            sub_params[t_param] = Parameter(new_param_name)
+                            t_param_exp = t_param_exp.assign(t_param, sub_params[t_param])
+                            sub_node_params.append(t_param_exp)
+                            template_params.append(t_param_exp)
+                        else:
+                            sub_node_params.append(t_param_exp)
+                            template_params.append(t_param_exp)
```

Comment:
These appends were at the wrong level - they only got added to the lists if the parameters were expressions.  That was a problem for `template_params` in particular, since it would have the wrong number of elements, and the subsequent zip between `circuit_params` and `template_params` would line things up incorrectly.

I added very explicit tests to show the failures in fc8315688ffb5ebca7e56c9b6f72887e90879dea, and fixed them in a14dc37441f73e8c9630d43b03ba8a1c9de5c757.